src/main/java/com/spotify/confidence/EventSenderEngineImpl.java

```diff
@@ -1,162 +1,131 @@
+import io.grpc.ManagedChannel;
 import java.io.IOException;
-import java.util.List;
-import java.util.Optional;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
 import java.util.concurrent.*;
-import org.slf4j.Logger;
+import java.util.concurrent.locks.LockSupport;
 
 class EventSenderEngineImpl implements EventSenderEngine {
+
   static final String EVENT_NAME_PREFIX = "eventDefinitions/";
-  private final ExecutorService writeThread = Executors.newSingleThreadExecutor();
-  private final ExecutorService uploadThread = Executors.newSingleThreadExecutor();
-  private final ConcurrentLinkedQueue<Event> writeQueue = new ConcurrentLinkedQueue<>();
-  private final BlockingQueue<String> uploadQueue = new LinkedBlockingQueue<>();
-  private final BlockingQueue<String> shutdownQueue = new LinkedBlockingQueue<>(1);
-  private final EventSenderStorage eventStorage = new InMemoryStorage();
+  static final int DEFAULT_BATCH_SIZE = 25;
+  static final Duration DEFAULT_MAX_FLUSH_INTERVAL = Duration.ofSeconds(60);
   private final EventUploader eventUploader;
-  private final int maxBatchSize;
   private final Clock clock;
-  private static final String UPLOAD_SIG = "UPLOAD";
-  private static final String SHUTDOWN_UPLOAD = "SHUTDOWN_UPLOAD";
-  private static final String SHUTDOWN_UPLOAD_COMPLETED = "SHUTDOWN_UPLOAD_COMPLETED";
-  private static final String SHUTDOWN_WRITE_COMPLETED = "SHUTDOWN_WRITE_COMPLETED";
-  private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
-  private final int flushTimeoutMilliseconds;
-  private ScheduledFuture<?> pendingFlush;
-  private volatile boolean isStopped = false;
-
-  private static final Logger log = org.slf4j.LoggerFactory.getLogger(EventSenderEngineImpl.class);
+  private final int maxBatchSize;
+  private final Duration maxFlushInterval;
+  private final FailsafeExecutor<Boolean> uploadExecutor;
+  private final ConcurrentLinkedQueue<Event> sendQueue = new ConcurrentLinkedQueue<>();
+  private final Set<CompletableFuture<?>> pendingBatches = ConcurrentHashMap.newKeySet();
+  private final Thread pollingThread = new Thread(this::pollLoop);
+  private volatile boolean intakeClosed = false;
 
+  @VisibleForTesting
   EventSenderEngineImpl(
-      int maxBatchSize, EventUploader eventUploader, Clock clock, int flushTimeoutMilliseconds) {
-    this.maxBatchSize = maxBatchSize;
+      int maxBatchSize, EventUploader eventUploader, Clock clock, Duration maxFlushInterval) {
     this.eventUploader = eventUploader;
     this.clock = clock;
-    this.flushTimeoutMilliseconds = flushTimeoutMilliseconds;
-    writeThread.submit(new WritePoller());
-    uploadThread.submit(new UploadPoller());
+    this.maxBatchSize = maxBatchSize;
+    this.maxFlushInterval = maxFlushInterval;
+    uploadExecutor =
+        Failsafe.with(
+            RetryPolicy.<Boolean>builder()
+                .handleResult(false)
```

Comment:
Need to verify is this (upload function returning `false`) is actually a valid reason to retry.

Reply:
So it only returns false on exceptions from grpc. Are there exceptions we don't want to retry?

Reply:
Ok, I'm going through the list of possible grpc errors and will make commit a proposal.